the-collab-lab/tcl-66-smart-shopping-list
# 21. Share/Unshare Feature

## Description

- Updated `modal `component to be reusable
- Created a `SharedWithList` component that allows users to see what users that have shared a list with
    - Each user shared with also has a delete button so that users can unshare
    - There is a multi-select / search tool for users to choose multiple lists and also search their lists
    - It will not show lists the user does not own, they can only share their own lists
 - InviteForm allows user to enter a name to attach to the user they are sharing it with so they can better identify later
 - Added a `useSharedWithData` and `unshareList` to the firebase file, and also updated `shareList` function to check if a list has already been shared with a user, and also update the list to have an array of objects that holds the users name and email
 - Decoupled the `capitalize` function to be more reusable

## Related Issue

closes #48 

## Acceptance Criteria

- [x] Create a SharedWithList component
- [x] Create logic that allows user to unshare a list they have shared
- [x] Create a button that shows how many users a list has been shared with
- [x] Update styling to match app/figma file


## Type of Changes

enhancement

## Updates

### Before
<img width="876" alt="Screenshot 2024-03-29 at 9 39 25 PM" src="https://github.com/the-collab-lab/tcl-66-smart-shopping-list/assets/98231995/934ef65d-c301-4b6e-8b77-4872f78630d6">

### After
<img width="879" alt="Screenshot 2024-03-29 at 9 39 47 PM" src="https://github.com/the-collab-lab/tcl-66-smart-shopping-list/assets/98231995/901e8c30-c982-4c6d-93e6-7b06402d938e">
<img width="879" alt="Screenshot 2024-03-29 at 9 40 00 PM" src="https://github.com/the-collab-lab/tcl-66-smart-shopping-list/assets/98231995/3206e999-f1f9-41d8-8746-8ce9fc6ba43c">
<img width="878" alt="Screenshot 2024-03-29 at 9 41 17 PM" src="https://github.com/the-collab-lab/tcl-66-smart-shopping-list/assets/98231995/b7b7a0b4-86e4-4481-9cc7-a8f47ae86f2f">


## Testing Steps / QA Criteria

1. git pull from main
2. git checkout an-share-feature
3. npm install
4. navigate to list page
5. if the list is owned by you, a share button should be available
6. share with a user (can use my email to test eonflower.dev@gmail.com)
7. once a list has been shared, it should pop up the icon with a number corresponding to the number of people its been shared with

Some notes as always:
- Previous shared lists will not have the proper schema since I updated lists to gain an array of users (their name and email), so they will not appear in the number with the icon
- I did light styling for shared with page, it wasn't really my focus, but am very open to changing it
- I personally liked having the share button on the list page instead of on the listItem, but we could decouple and make a component for it so that it can be used in multiple places
- I used the [react-select library ](https://react-select.com/home) for the select tool because this was how I envisioned it. I liked it, but also I didn't really ask, so we can remove it if y'all don't like it
